shared/js/passcode_helper.js

```diff
@@ -0,0 +1,147 @@
+      });
+    },
+    setPasscode: function(newPass) {
+      var lock = navigator.mozSettings.createLock();
+      var getFromSettings = Promise.all([
+        lock.get(SET_DIGEST_ITERATIONS),
+        lock.get(SET_DIGEST_ALGORITHM)
+      ]);
+      var digest = getFromSettings.then((values) => {
+        // Always generate a new salt.
+        this.salt = crypto.getRandomValues(new Uint8Array(8));
+        /* the combined lock.get() makes it quite ugly. we get
+         an Array of objects, each with just one key/value,
+         which is the requested setting. let's destruct as follows:
+         */
+        this.iterations = parseInt(values[0][SET_DIGEST_ITERATIONS]) ||
+          DEFAULT_ITERATIONS;
+        this.algorithm = (typeof(values[1][SET_DIGEST_ALGORITHM]) == 'string') ?
+          values[1][SET_DIGEST_ALGORITHM] : DEFAULT_ALGORITHM;
+        var self = this;
+        return this._makeDigest(newPass, this.salt,
+                        this.iterations, this.algorithm)
+          .then(self._storeNewDigest.bind(self)).catch((error) => {
+            console.error('PasscodeHelper: Could not make digest:', error);
+            return Promise.resolve(false);
+          });
+      }).catch((error) => {
+        console.error('PasscodeHelper: No Settings?', error);
+        return Promise.resolve(false);
+        });
+      return digest;
+    },
+    _storeNewDigest: function (digest) {
+    // Note: We can now store the salt, since digest was generated!
+    var digestUint8 = new Uint8Array(digest);
+    var newSettings = {};
+    newSettings[SET_DIGEST_SALT] = this.salt;
+    newSettings[SET_DIGEST_VALUE] = digestUint8;
+    newSettings[SET_DIGEST_ITERATIONS] = this.iterations;
+    newSettings[SET_DIGEST_ALGORITHM] = this.algorithm;
+    var lock = navigator.mozSettings.createLock();
+    return lock.set(newSettings).then(() => {
+      return digestUint8;
+    }).catch((error) => {
+      console.error('PasscodeHelper: Couldnt store new digest');
+      return Promise.resolve(false);
+    });
+  },
+    checkPasscode: function (testPass) {
+      this.testPass = testPass;
+      //get salt & digest out of settings
+      var lock = navigator.mozSettings.createLock();
+      var storedParams = Promise.all([
+        lock.get(SET_DIGEST_SALT),
+        lock.get(SET_DIGEST_ITERATIONS),
+        lock.get(SET_DIGEST_ALGORITHM),
+        lock.get(SET_DIGEST_VALUE)
+      ]);
+      return storedParams.then(this._makeAndCompare.bind(this))
+      .catch((error) => {
+        console.error('PasscodeHelper: Couldnt get digest Settings:', error);
+        return Promise.resolve(false);
+      });
+    },
+    _makeAndCompare: function(values) {
+      /* the combined lock.get() makes it quite ugly. we get
+       an Array of objects, each with just one key/value,
+       which is the requested setting. let's destruct as follows:
+       */
+      var salt = this._toTypedArray(values[0][SET_DIGEST_SALT]);
```

Comment:
The ugly format like `[{setting1: settings1value}, {setting2, setting2value}, ..]`can't be avoided due to the settings API.

The question that remains is where to destruct?
Should I rewrite line 107 to something like this?
```js
      return storedParams.then((values) => {
        //XX destruct here
        this._makeAndCompare();
      })
```